feat(withTrackingComponentDecorator): rewrite context behavior to improve performance

BRKalow opened the pull request:
The React 16 context has slightly different behavior when dealing with updates in a subtree. With the legacy context, a `shouldComponentUpdate` returning false in a component tree would cause the context consumers in its subtree to not receive context updates. From the React docs:

> The problem is, if a context value provided by component changes, descendants that use that value won’t update if an intermediate parent returns false from shouldComponentUpdate.
-- https://reactjs.org/docs/legacy-context.html#updating-context

With the new context introduced in React 16, the context consumers will always update regardless of a parent's optimizations:
> All consumers that are descendants of a Provider will re-render whenever the Provider’s value prop changes. The propagation from Provider to its descendant consumers is not subject to the shouldComponentUpdate method, so the consumer is updated even when an ancestor component bails out of the update.
-- https://reactjs.org/docs/context.html#contextprovider

What we encountered within our application was that after the upgrade we were seeing significantly more re-renders of our component trees than before, causing some performance issues. 

We were able to fix the re-render regressions by rewriting how the context is structured and how data is merged in a tree of components decorated with `track`.

Instead of eagerly computing the tracking context data, we are providing a callback on the context that will merge the data bottom-up. We also changed the dispatch function to use a similar mechanism. This has the benefit of the context value having stable referential equality, so there are no extra renders caused by the context provider updating with the eagerly computed tracking data.

We forked `react-tracking` internally and made these updates, which we have been using for a few weeks. These changes brought our performance metrics back to where they were before the `react-tracking@6` upgrade. We also added a test case in the `e2e` test suite which represents the expected behavior (the test case fails on the current master branch).

This technically changes the context shape as we removed `context.tracking.data` and replaced it with `context.tracking.getTrackingData()`, so it will require a major version bump.

**Simple reproduction case:**
6.0.0 behavior: https://codesandbox.io/s/frosty-bouman-7hfmr
5.7.0 behavior: https://codesandbox.io/s/mmq0zn57jp

Thanks!

_Co-authored with: @jacekradko_

tizmagik commented:
Thank you for this, awesome PR. Thank you especially for the additional test case. 

I’m almost inclined not to bump semver major since folks aren’t supposed to be interacting with the context shape directly, but instead rely on the interface of the `props.tracking` object shape. Will think about that one for release -- curious to hear your thoughts on this as well.

tizmagik commented on src/__tests__/e2e.test.js:
```diff
@@ -585,6 +587,51 @@ describe('e2e', () => {
+    });
+
+    const App = track()(() => {
+      const [props, setProps] = React.useState({});
```
I know these can be called anything you want, but by [convention/docs](https://reactjs.org/docs/hooks-reference.html#usestate) these are usually:

```js
const [state, setState] = React.useState({});
```

tizmagik commented on src/useTracking.js:
```diff
@@ -15,12 +14,12 @@ export default function useTracking() {
-          merge(trackingContext.tracking.data, data)
-        ),
+      getTrackingData: () => trackingContext.tracking.getTrackingData(),
+      trackEvent: data => trackingContext.tracking.dispatch(data),
```
Would it make sense to avoid the extra function declaration?

```js
      getTrackingData: trackingContext.tracking.getTrackingData,
      trackEvent: trackingContext.tracking.dispatch,
````

BRKalow replied:
Good call!

BRKalow commented:
@tizmagik Thanks for the quick review! I've addressed all your comments.

Regarding whether or not to bump a major version, my first thought was also the same as yours. Since the context shape is intended to only be used internally and it is not a public, documented API it would be acceptable to leave this change as a minor version bump. 👍

...maybe the context type needs to be exported as `__INTERNAL_DO_NOT_USE_ReactTrackingContext` 😄

tizmagik commented:
Actually I'm wondering if we should be exporting the context type at all? I'm not sure there's a use case for that since the hook is importable as `{ useTracking }` ... 🤔

BRKalow commented:
@tizmagik Thats a good point. With the new hook, it should provide the methods needed to make additional custom hooks. It would reduce the public API surface and perhaps give more flexibility going forward for updating the internals.

tizmagik commented:
Thanks, I'll merge this and then submit a separate PR to remove the context type export since I don't think we want folks to be using that interface at all. Or, if there is a use case, we can follow-up with a separate issue/PR in the future.

tizmagik commented:
This was released in 7.1.0 https://github.com/nytimes/react-tracking/releases/tag/v7.1.0
